### PR TITLE
Sync PrepareProposal with Spec. Main part

### PR DESCRIPTION
**Pull request by sergio-mena**
This is the core of the logic update for `PrepareProposal` to match the specification.
* The main changes are coming from #8094. PR #8094 also included three extra "semi-related yak-shaves":
  * `CreateProposalBlock` no longer returns the block's `PartSet`. **This has been kept**
  * Using a `mockery` type for the mempool, replacing the hand written one. **This has been kept**
  * Introduction of library function to hash the ResponseFinalizeBlock.ExecTxResults, and related refactoring. **This has been dropped**, as it depends on `FinalizeBlock`. Will probably be useful when backporting `FinalizeBlock`-related code
* There were a few other v0.35.x/v0.36.x PRs that I needed to backport in order to have the needed functionality
  * Partial backport of #6885. The PR is related to vote extensions, but introduces some `PrepareProposal` logic in the `persistent_kvstore` application
  * Full backport of #6567 and #7501, which contain some test logic refactoring used everywhere
  * Partial backport of #7961. PR #7961 implements `ProcessProposal`, but there were two infra improvements in it that we needed here
    * EvidenceList's method ToABCI
    * abci App mocks

---

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

**Comment by sergio-mena** [comment]
Addresses #9118

File: consensus/state.go [hunk 1]
```diff
@@ -1131,8 +1131,17 @@ func (cs *State) defaultDecideProposal(height int64, round int32) {
...
-		block, blockParts = cs.createProposalBlock()
-		if block == nil {
+		var err error
+		block, err = cs.createProposalBlock()
+		if err != nil {
+			cs.Logger.Error("unable to create proposal block", "error", err)
+			return
+		} else if block == nil {
+			return
+		}
```
Comment (cmwaters) [inline consensus/state.go:1141]: When does this happen?

File: state/execution.go [hunk 2]
```diff
@@ -128,19 +134,28 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
...
-	return state.MakeBlock(height, modifiedTxs, commit, evidence, proposerAddr)
+	for _, rtx := range txrSet.RemovedTxs() {
+		if err := blockExec.mempool.RemoveTxByKey(rtx.Key()); err != nil {
+			blockExec.logger.Debug("error removing transaction from the mempool", "error", err, "tx hash", rtx.Hash())
+		}
+	}
+	for _, atx := range txrSet.AddedTxs() {
+		if err := blockExec.mempool.CheckTx(atx, nil, mempool.TxInfo{}); err != nil {
+			blockExec.logger.Error("error adding tx to the mempool", "error", err, "tx hash", atx.Hash())
+		}
```
Comment (cmwaters) [inline state/execution.go:155]: I didn't know we did that. But why do we need to run `CheckTx` again?

File: types/tx.go [hunk 3]
```diff
@@ -105,13 +107,182 @@ func (txs Txs) ToSliceOfBytes() [][]byte {
...
+// TxRecordSet contains indexes into an underlying set of transactions.
+// These indexes are useful for validating and working with a list of TxRecords
+// from the PrepareProposal response.
+//
+// Only one copy of the original data is referenced by all of the indexes but a
+// transaction may appear in multiple indexes.
+type TxRecordSet struct {
+	// all holds the complete list of all transactions from the original list of
+	// TxRecords.
+	all Txs
```
Comment (cmwaters) [inline types/tx.go:119]: Is this necessary? Doesn't look like we use it

**Review by cmwaters — APPROVED** [review]

File: state/execution.go (hunk 2 above)
Comment (williambanfield) [reply state/execution.go:155]: These are transactions marked as `Added`, I _believe_ the semantics at the time were that Tx marked added are placed into the mempool. Was then removed when we thought better of this idea in https://github.com/tendermint/tendermint/pull/8176

File: types/tx.go (hunk 3 above)
Comment (williambanfield) [reply types/tx.go:119]: It's used in the validation logic to check for duplicates.

**Review by williambanfield — APPROVED** [review]

File: state/execution.go (hunk 2 above)
Comment (cmwaters) [reply state/execution.go:155]: Yeah Sergio followed up with me async. Thanks for the context

File: consensus/state.go (hunk 1 above)
Comment (sergio-mena) [reply consensus/state.go:1141]: Good point. I checked the new logic of `createProposalBlock` and it seems it's not possible (either you get the block or an error).
I'll change the return by a panic.

File: state/execution.go (hunk 2 above)
Comment (sergio-mena) [reply state/execution.go:155]: Yeah, as soon as I get this PR merges I'll post the last PR, which contains #8176, which was fixing this on v0.36.x

File: consensus/state.go (hunk 1 above)
Comment (sergio-mena) [reply consensus/state.go:1141]: BTW, this is something I didn't catch as a reviewer last time (in #8094) 😉

**Pull request merged**